pkg/issuer/acme/dns/util/wait.go

```diff
@@ -66,32 +70,42 @@ func getNameservers(path string, defaults []string) []string {
 	return systemNameservers
 }
 
-// Update FQDN with CNAME if any
-func updateDomainWithCName(r *dns.Msg, fqdn string) string {
-	for _, rr := range r.Answer {
-		if cn, ok := rr.(*dns.CNAME); ok {
-			if cn.Hdr.Name == fqdn {
-				logf.V(logf.DebugLevel).Infof("Updating FQDN: %s with its CNAME: %s", fqdn, cn.Target)
-				fqdn = cn.Target
-				break
+// Update FQDN with CNAME if any, it will follow CNAME records till it hits a non-CNAME.
+// this will error if there is a recursive CNAME in the chain
+func updateDomainWithCName(fqdn string, nameservers []string, fqdnChain ...string) (string, error) {
+	r, err := dnsQuery(fqdn, dns.TypeCNAME, nameservers, true)
+	if err == nil && r.Rcode == dns.RcodeSuccess {
+		for _, rr := range r.Answer {
+			if cn, ok := rr.(*dns.CNAME); ok {
+				if cn.Hdr.Name == fqdn {
+					logf.V(logf.DebugLevel).Infof("Updating FQDN: %s with its CNAME: %s", fqdn, cn.Target)
+
+					// check if we were here before to prevent recursive records causing issues
+					for _, fqdnInChain := range fqdnChain {
+						if cn.Target == fqdnInChain {
+							return "", fmt.Errorf("Found recursive CNAME record to %q when looking up %q", cn.Target, fqdn)
+						}
+					}
+					return updateDomainWithCName(cn.Target, nameservers, append(fqdnChain, fqdn)...)
+				}
 			}
 		}
+	} else if err != nil {
+		return "", err
 	}
 
-	return fqdn
+	return fqdn, nil
 }
```

Comment:
I am a bit confused with the depth of this function 😅 I suggest that we keep the [happy path to the left](https://maelvls.dev/go-happy-line-of-sight/):

```suggestion
func updateDomainWithCName(fqdn string, nameservers []string, fqdnChain ...string) (string, error) {
	r, err := dnsQuery(fqdn, dns.TypeCNAME, nameservers, true)
	if err != nil {
		return "", err
	}
	if r.Rcode != dns.RcodeSuccess {
		return fqdn, err
	}

	for _, rr := range r.Answer {
		cn, ok := rr.(*dns.CNAME)
		if !ok || cn.Hdr.Name != fqdn {
			continue
		}

		logf.V(logf.DebugLevel).Infof("Updating FQDN: %s with its CNAME: %s", fqdn, cn.Target)

		// check if we were here before to prevent recursive records causing issues
		for _, fqdnInChain := range fqdnChain {
			if cn.Target != fqdnInChain {
				continue
			}

			return "", fmt.Errorf("Found recursive CNAME record to %q when looking up %q", cn.Target, fqdn)
		}

		return updateDomainWithCName(cn.Target, nameservers, append(fqdnChain, fqdn)...)
	}

	return fqdn, nil
}
```

Reply:
Looks better, i was honouring the old code a bit too much